ddl/backfilling_operator.go

```diff
@@ -0,0 +1,511 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/tidb/ddl/ingest"
+	"github.com/pingcap/tidb/ddl/internal/session"
+	"github.com/pingcap/tidb/disttask/operator"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/parser/terror"
+	"github.com/pingcap/tidb/resourcemanager/pool/workerpool"
+	"github.com/pingcap/tidb/resourcemanager/util"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	_ operator.Operator                = (*TableScanTaskSource)(nil)
+	_ operator.WithSink[TableScanTask] = (*TableScanTaskSource)(nil)
+
+	_ operator.WithSource[TableScanTask]  = (*TableScanOperator)(nil)
+	_ operator.Operator                   = (*TableScanOperator)(nil)
+	_ operator.WithSink[IndexRecordChunk] = (*TableScanOperator)(nil)
+
+	_ operator.WithSource[IndexRecordChunk] = (*IndexIngestOperator)(nil)
+	_ operator.Operator                     = (*IndexIngestOperator)(nil)
+	_ operator.WithSink[IndexWriteResult]   = (*IndexIngestOperator)(nil)
+
+	_ operator.WithSource[IndexWriteResult] = (*indexWriteResultSink)(nil)
+	_ operator.Operator                     = (*indexWriteResultSink)(nil)
+)
+
+type opSessPool interface {
+	Get() (sessionctx.Context, error)
+	Put(sessionctx.Context)
+}
+
+// NewAddIndexIngestPipeline creates a pipeline for adding index in ingest mode.
+func NewAddIndexIngestPipeline(
+	ctx context.Context,
+	store kv.Storage,
+	sessPool opSessPool,
+	engine ingest.Engine,
+	sessCtx sessionctx.Context,
+	tbl table.PhysicalTable,
+	idxInfo *model.IndexInfo,
+	startKey, endKey kv.Key,
+) (*operator.AsyncPipeline, error) {
+	index := tables.NewIndex(tbl.GetPhysicalID(), tbl.Meta(), idxInfo)
+	copCtx, err := NewCopContext(tbl.Meta(), idxInfo, sessCtx)
+	if err != nil {
+		return nil, err
+	}
+	poolSize := copReadChunkPoolSize()
+	srcChkPool := make(chan *chunk.Chunk, poolSize)
+	for i := 0; i < poolSize; i++ {
+		srcChkPool <- chunk.NewChunkWithCapacity(copCtx.fieldTps, copReadBatchSize())
+	}
+	readerCnt, writerCnt := expectedIngestWorkerCnt()
+
+	srcOp := NewTableScanTaskSource(ctx, store, tbl, startKey, endKey)
+	scanOp := NewTableScanOperator(ctx, sessPool, copCtx, srcChkPool, readerCnt)
+	ingestOp := NewIndexIngestOperator(ctx, copCtx, sessPool, tbl, index, engine, srcChkPool, writerCnt)
+	sinkOp := newIndexWriteResultSink(ctx)
```

Comment:
What will happen when create worker failed?

Reply:
I will add failure tests in next PR.